File: aoc/aoc15/python/day2.py
def readFile(filepath):
    with open(filepath, "r") as f:
        return [line.strip() for line in f.readlines()]

def solve(depths):
    totalPaper = 0
    totalRibbon = 0

    for x in depths:
        xyz = [int(el) for el in x.split("x")]
        face1 = xyz[0] * xyz[1]
        face2 = xyz[0] * xyz[2]
        face3 = xyz[1] * xyz[2]

        minFace = min(face1, face2, face3)

        paperRequired = 2 * (face1 + face2 + face3) + minFace
    
        totalPaper += paperRequired

        maxSide = max(xyz[0], xyz[1], xyz[2])
        perim = 2 * (xyz[0] + xyz[1] + xyz[2] - maxSide)
        ribbonRequired = perim + xyz[0] * xyz[1] * xyz[2]
        totalRibbon += ribbonRequired
        

    return (totalPaper, totalRibbon)

depths = readFile('./inputs/day2.txt')
solution = solve(depths)
print("Total Paper = ", solution[0], ", total Ribbon = ", solution[1])

File: aoc/aoc15/rust/src/bin/day2.rs
use std::fs;

fn main()
{
    let input = fs::read_to_string("../inputs/day2.txt").expect("Failed to read input");

    let result = run(input);
    println!("{}, {}", result.0, result.1);
}

fn run(input: String) -> (i32, i32)
{
    let lines = input.lines();
    let mut total = (0, 0);
    for line in lines
    {
        let lval = run_line(line.to_string());
        total.0 += lval.0;
        total.1 += lval.1;
    }
    total
}

fn run_line(input: String) -> (i32, i32)
{
    let mut parts = input.split("x");
    let x: i32= parts.next().expect("x").parse().unwrap();
    let y: i32= parts.next().expect("y").parse().unwrap();
    let z: i32= parts.next().expect("z").parse().unwrap();

    let s1 = x*y;
    let s2 = x*z;
    let s3 = y*z;

    let min = s1.min(s2).min(s3);

    (2 * (s1 + s2 + s3) + min, calc_ribbon(x, y, z))
}

fn calc_ribbon(x: i32, y: i32, z: i32) -> i32
{
    let max = x.max(y).max(z);
    let perim = 2 * (x + y + z - max);
    perim + x * y * z
}

#[test]
fn example()
{
    assert_eq!(58, run_line("2x3x4".to_string()).0);
    assert_eq!(34, run_line("2x3x4".to_string()).1);
    assert_eq!(58, run("2x3x4".to_string()).0);
}